### ISSUE TITLE
Client doesn't notice signal while in waiting loop

### ISSUE THREAD

**Issue by Sebastian-Roth** [FOGProject/fog-client#93]
From what it looks like the fog-client does not shutdown in an appropriate time when being requested (e.g. on system shutdown). This has been tested on Ubuntu 16.04.1 so far but would most probably apply to other linux OS as well.

/opt/fog-service/control.sh sends `kill -15` but probably the client does not see this signal while in the waiting loop.

**Comment by Sebastian-Roth** [comment]
Reference for more information on this: https://forums.fogproject.org/topic/10811/fog-client-on-ubuntu-16

**Comment by FinancedHydra** [comment]
I can think of two possibilities:

1) `mono-service` has a bug (unlikely)
2) `control.sh` has become de-attached from the FOGService process; the `service.lock` file would have to contain the wrong `pid`

Some background:

The FOG Client is actually being wrapped by `mono-service`, which enables c# services to run on unix and accept sigterm/kill signals, as if they were `net stop`. Those signal handlers are not blocked by the module thread waiting. Zazzles actually puts all the client work on new thread as to avoid possible signal blockage: https://github.com/FOGProject/zazzles/blob/master/Zazzles/AbstractService.cs#L69

**Comment by Sebastian-Roth** [comment]
I have looked into this again as I was hoping to be able to find and fix this for the upcoming release. On Ubuntu 16.04 the issue seems easy to replicate. Install FOGService and after another reboot you see it hang for at least 30 seconds, sometimes longer. Seems like it can be replicated when simply running `systemctl stop FOGService` - this command hangs as well.

Interesting things I noted:
* When installing FOGService at first it was started right away and running properly (checked in fog.log) but didn't cause a hang on shutdown/reboot the machine. Following reboots always caused the hang.
* It seems like some of the service is being shut down quickly as we see FOGUserService.exe is gone from the process list (`ps ax | grep -i fog | grep -v grep`) but `... mono-service.exe ... FOGService.exe` is still running.

Next step would be to test on other Linux systems to see if this is a bug in mono-service and as well add seconds to `fog.log` to be able to see if FOGService does terminate early or not. Right now it logs minutes only and this doesn't give enough information.

**Comment by Sebastian-Roth** [comment]
After some extensive debugging sessions and digging through many `strace` outputs I figured this seems to be caused by the external library [Quartz](https://github.com/quartznet/quartznet) we use to schedule tasks. Commenting line 46 and 47 in [PowerManagement class](https://github.com/FOGProject/fog-client/blob/master/Modules/PowerManagement/PowerManagement.cs#L46) does fix the hang on shutdown problem. Tested on Ubuntu 20.04 (pre release) so far but I am fairly sure this should be the case in other Linux systems as well.

We currently still use Quartz 2.4.1 because the newer 3.x versions have a dependency to .Net 4.5.2 at least. Updating to Quartz 2.6.2 (the last of the 2.x branch) does not fix the issue. Testing 3.x is still on the todo list.

**Comment by Sebastian-Roth** [comment]
Another report in the forums: https://forums.fogproject.org/topic/14348/fog-client-on-ubuntu-18-04

**Comment by Sebastian-Roth** [comment]
Finally got around to test upgrading our project to 4.5.2 (thanks to @danielbmiller!) to be able to use Quartz 3.x. The mentioned issue with shutdown hanging seems to be fixed in Quartz 3.x!

Will need more testing on Windows to make sure this is not causing other issues.

- [x] Test shutdown (no hang) on Ubuntu pre-20.04
- [x] Test task reboot and PowerManagement cron style reboot task on Linux
- [ ] Test shutdown (no hang) on other Linux distro ([jc35](https://forums.fogproject.org/topic/14348/fog-client-on-ubuntu-18-04))
- [ ] Test general behavior on Windows with .Net 4.5.2 and Quartz 3.x 
- [ ] Test task reboot and PowerManagement cron style reboot task on Windows

**Comment by florentlagoda** [comment]
Hi,
I can test on Windows 10 if you want.
How to test ?

**Comment by Sebastian-Roth** [comment]
@florentlagoda Thanks! While this particular issue mentioned here is not specific to Windows we had to move to a newer version of .Net and Quartz to get around this and it would be great to get this build tested on Windows as well before we go ahead.

The changes haven't been committed to the repo yet. So the only way to test for now is using the SmartInstaller.exe provided here: https://github.com/FOGProject/fog-client/releases/download/0.11.19/SmartInstaller_Linux_stop_hang_fixed.exe

Be aware this installer is not meant to be used in production as its binaries are not signed (needed for auto updating). Please use this for testing on one (or a few at most) client/s to see if it works properly. Please test scheduling a deploy task (shutdown GUI, reboot) as well as Power Management tasks (reboot, shutdown) to make sure the new Quartz library works as intended. Thanks in advance!

**Comment by jc35r2** [comment]
Hi, I can test on xubuntu 18.04 et 20.04.
How to test ?

**Comment by Sebastian-Roth** [comment]
@jc35r2 As mentioned [in the forums](https://forums.fogproject.org/topic/14348/fog-client-on-ubuntu-18-04) you can download the [fixed SmartInstaller](https://github.com/FOGProject/fog-client/releases/download/0.11.19/SmartInstaller_Linux_stop_hang_fixed.exe). Install that on Linux and do a couple of reboots to see if it still hangs on shutdown.

If you've got an older version still installed you can run `mono SmartInstaller.exe --uninstall` to remove the old install first.

**Comment by Sebastian-Roth** [comment]
@florentlagoda @jc35r2 Any news on the testing? Thanks in advance!

**Comment by jc35r2** [comment]
Hi,
Sorry for a late response.
I did a test but the client does not work at all.
I have a password timeout message.
Because of the confinement, we work remotely, I cannot provide you with the logs. We lost control of our testing machines. As soon as we can access our office, we can restart our machines. I still tried to launch a snapin and it didn't work.

**Comment by florentlagoda** [comment]
Hi, @Sebastian-Roth ,
It's ok on Windows 10 x64. I have only test the reboot task (due to teleworking) and postpone once

**Comment by Sebastian-Roth** [comment]
@jc35r2 I am not sure I understand this. How were you able to install the new fog-client provided for testing. Did you have SSH access to the machine before and now it's gone? What do you mean by "I have a password timeout message"? SSH password login timeout?

@florentlagoda Thanks as well! Good to hear it seems to work on Windows 10!

**Comment by jc35r2** [comment]
"password timeout" I have this in the fog.log on the client. I wanted to show the fog.log here, but I can't access to the machine copy the fog.log.

**Comment by Sebastian-Roth** [comment]
@jc35r2 "password timeout" doesn't ring a bell for me. Not that I know the code from start to end but I think I haven't come across this one yet. We'll need the log then.

How were you able to login to the machine before? Via SSH? Doesn't work anymore? Does it respond to ping?

**Comment by jc35r2** [comment]
Hi,
I install SmartInstaller_Linux_stop_hang_fixed.exe
mono /opt/SmartInstaller_Linux_stop_hang_fixed.exe --server="fogserver" --webroot="/fog" -s -t 0

The log is on https://forums.fogproject.org/topic/14348/fog-client-on-ubuntu-18-04/5
Because here, I have a message 'You can't comment at this time.'

**Comment by jc35r2** [comment]
I try 2 machines and it's OK.

**Comment by jc35r2** [comment]
Hi,
Yesterday I have the same problem on a machine, when a powerdown.
I retry this. The client can't contact the server because the vpn is down.
